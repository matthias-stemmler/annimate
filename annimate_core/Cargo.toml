[package]
name = "annimate_core"
version = "0.1.0"
edition.workspace = true

[build-dependencies]
cargo_metadata = "0.18"

[dependencies]
csv = "1.0"
graphannis = "3.2.1"
graphannis-core = "3.2.1"
itertools = { workspace = true }
regex = { workspace = true }
serde = { workspace = true }
serde_with = "3.7"
thiserror = { workspace = true }
toml = "0.8"
urlencoding = "2.1"

[dev-dependencies]
indoc = "2.0"
insta = { version = "1.34", features = ["serde"] }
