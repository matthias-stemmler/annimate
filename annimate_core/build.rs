//! Build script for `annimate_core`.

use cargo_metadata::MetadataCommand;

fn main() {
    inject_graphannis_version();
}

fn inject_graphannis_version() {
    let metadata = MetadataCommand::new()
        .exec()
        .expect("Failed to run `cargo metadata`");

    let graphannis_version = metadata
        .packages
        .into_iter()
        .find(|p| p.name == "graphannis")
        .expect("`graphannis` package not found in Cargo metadata`")
        .version
        .to_string();

    println!("cargo::rustc-env=GRAPHANNIS_VERSION={graphannis_version}");
}
