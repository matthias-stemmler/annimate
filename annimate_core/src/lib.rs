use std::io::Write;
use std::path::{Path, PathBuf};

use anno::AnnoKeys;
use corpus::CorpusRef;
use format::export;
use graphannis::corpusstorage::CacheStrategy;
use import::{FilesystemEntity, ImportFormat, ImportableCorpus};
use itertools::Itertools;
use metadata::{CorpusSet, MetadataStorage};
use query::Query;
use serde::Serialize;

mod anno;
mod aql;
mod corpus;
mod error;
mod format;
mod import;
mod metadata;
mod node_name;
mod query;
mod util;
mod version;

pub use anno::{ExportableAnnoKey, ExportableAnnoKeys};
pub use aql::{QueryAnalysisResult, QueryNode, QueryNodes};
pub use error::AnnisExportError;
pub use format::{CsvExportColumn, CsvExportConfig, ExportFormat};
pub use graphannis::corpusstorage::CorpusInfo;
pub use graphannis::graph::AnnoKey;
pub use import::ImportedCorpus;
pub use query::{ExportData, ExportDataAnno, ExportDataText, QueryLanguage};
pub use version::{VersionInfo, VERSION_INFO};

pub struct Storage {
    corpus_storage: graphannis::CorpusStorage,
    metadata_storage: MetadataStorage,
}

impl Storage {
    pub fn from_db_dir<P>(db_dir: P) -> Result<Self, AnnisExportError>
    where
        P: AsRef<Path>,
    {
        let corpus_storage = graphannis::CorpusStorage::with_cache_strategy(
            db_dir.as_ref(),
            CacheStrategy::PercentOfFreeMemory(25.0),
            true,
        )?;

        let metadata_storage = MetadataStorage::from_db_dir(
            &db_dir,
            &corpus_storage
                .list()?
                .into_iter()
                .map(|c| c.name)
                .collect_vec(),
        )?;

        Ok(Self {
            corpus_storage,
            metadata_storage,
        })
    }

    pub fn corpora(&self) -> Result<Corpora, AnnisExportError> {
        let sets = self.metadata_storage.corpus_sets();

        let corpora = self
            .corpus_storage
            .list()?
            .into_iter()
            .sorted_by(|a, b| a.name.cmp(&b.name))
            .map(|c| {
                let included_in_sets = sets
                    .iter()
                    .filter(|(_, CorpusSet { corpus_names })| corpus_names.contains(&c.name))
                    .map(|(name, _)| name.clone())
                    .collect();

                Corpus {
                    name: c.name,
                    included_in_sets,
                }
            })
            .collect();

        Ok(Corpora {
            sets: sets.keys().cloned().collect(),
            corpora,
        })
    }

    pub fn delete_corpus(&self, corpus_name: &str) -> Result<(), AnnisExportError> {
        self.corpus_storage.delete(corpus_name)?;
        self.metadata_storage.update_corpus_sets(|corpus_sets| {
            for CorpusSet { corpus_names } in corpus_sets.values_mut() {
                corpus_names.remove(corpus_name);
            }
        })?;
        Ok(())
    }

    pub fn import_corpora<F>(
        &self,
        paths: Vec<PathBuf>,
        on_status: F,
    ) -> Result<Vec<String>, AnnisExportError>
    where
        F: Fn(ImportStatusEvent),
    {
        let mut imported_corpus_names = Vec::new();

        let importable_corpora = import::find_importable_corpora(paths, |message| {
            on_status(ImportStatusEvent::Message {
                index: None,
                message: message.into(),
            });
        })?;

        on_status(ImportStatusEvent::CorporaFound {
            corpora: importable_corpora.iter().map_into().collect(),
        });

        for (index, corpus) in importable_corpora.into_iter().enumerate() {
            on_status(ImportStatusEvent::CorpusImportStarted { index });

            let result = import::import_corpus(&self.corpus_storage, corpus, |message| {
                on_status(ImportStatusEvent::Message {
                    index: Some(index),
                    message: message.into(),
                });
            });

            match result {
                Ok(imported_corpus) => {
                    imported_corpus_names.push(imported_corpus.imported_name.clone());

                    on_status(ImportStatusEvent::CorpusImportFinished {
                        index,
                        result: ImportCorpusResult::Imported {
                            corpus: imported_corpus,
                        },
                    });
                }
                Err(err) => {
                    on_status(ImportStatusEvent::CorpusImportFinished {
                        index,
                        result: ImportCorpusResult::Failed {
                            message: err.to_string(),
                        },
                    });
                }
            }
        }

        Ok(imported_corpus_names)
    }

    pub fn add_corpora_to_set<S>(
        &self,
        corpus_set_name: &str,
        corpus_names: &[S],
    ) -> Result<(), AnnisExportError>
    where
        S: AsRef<str>,
    {
        if corpus_names.is_empty() {
            return Ok(());
        }

        self.metadata_storage.update_corpus_sets(|corpus_sets| {
            corpus_sets
                .entry(corpus_set_name.into())
                .or_default()
                .corpus_names
                .extend(corpus_names.iter().map(|c| c.as_ref().into()));
        })?;

        Ok(())
    }

    pub fn toggle_corpus_in_set(
        &self,
        corpus_set_name: &str,
        corpus_name: &str,
    ) -> Result<(), AnnisExportError> {
        self.metadata_storage.update_corpus_sets(|corpus_sets| {
            if let Some(CorpusSet { corpus_names }) = corpus_sets.get_mut(corpus_set_name) {
                if !corpus_names.remove(corpus_name) {
                    corpus_names.insert(corpus_name.into());
                }
            }
        })?;

        Ok(())
    }

    pub fn validate_query<S>(
        &self,
        corpus_names: &[S],
        aql_query: &str,
        query_language: QueryLanguage,
    ) -> Result<QueryAnalysisResult<()>, AnnisExportError>
    where
        S: AsRef<str>,
    {
        Ok(aql::validate_query(
            self.corpus_ref(corpus_names),
            aql_query,
            query_language,
        )?)
    }

    pub fn query_nodes(
        &self,
        aql_query: &str,
        query_language: QueryLanguage,
    ) -> Result<QueryAnalysisResult<QueryNodes>, AnnisExportError> {
        Ok(aql::query_nodes(
            &self.corpus_storage,
            aql_query,
            query_language,
        )?)
    }

    pub fn exportable_anno_keys<S>(
        &self,
        corpus_names: &[S],
    ) -> Result<ExportableAnnoKeys, AnnisExportError>
    where
        S: AsRef<str>,
    {
        Ok(AnnoKeys::new(self.corpus_ref(corpus_names)).map(AnnoKeys::into_exportable)?)
    }

    pub fn segmentations<S>(&self, corpus_names: &[S]) -> Result<Vec<String>, AnnisExportError>
    where
        S: AsRef<str>,
    {
        Ok(anno::segmentations(self.corpus_ref(corpus_names))?)
    }

    pub fn export_matches<F, S, W>(
        &self,
        corpus_names: &[S],
        aql_query: &str,
        query_language: QueryLanguage,
        format: ExportFormat,
        mut out: W,
        on_status: F,
    ) -> Result<(), AnnisExportError>
    where
        F: Fn(ExportStatusEvent),
        S: AsRef<str>,
        W: Write,
    {
        let anno_keys = AnnoKeys::new(self.corpus_ref(corpus_names))?;
        let query = Query::new(self.corpus_ref(corpus_names), aql_query, query_language)?;
        let matches = query.find(format.get_export_data().cloned())?;

        on_status(ExportStatusEvent::Found {
            count: matches.total_count(),
        });

        export(
            format,
            matches,
            query.nodes(),
            anno_keys.format(),
            &mut out,
            |progress| on_status(ExportStatusEvent::Exported { progress }),
        )?;

        out.flush()?;

        Ok(())
    }

    fn corpus_ref<'a, S>(&'a self, corpus_names: &'a [S]) -> CorpusRef<'a, S> {
        CorpusRef::new(&self.corpus_storage, corpus_names)
    }
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Corpora {
    corpora: Vec<Corpus>,
    sets: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Corpus {
    name: String,
    included_in_sets: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum ExportStatusEvent {
    Found { count: usize },
    Exported { progress: f32 },
}

#[derive(Clone, Debug, Serialize)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum ImportStatusEvent {
    CorporaFound {
        corpora: Vec<ImportCorpus>,
    },
    CorpusImportStarted {
        index: usize,
    },
    CorpusImportFinished {
        index: usize,
        result: ImportCorpusResult,
    },
    Message {
        index: Option<usize>,
        message: String,
    },
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportCorpus {
    file_name: String,
    format: ImportFormat,
    trace: Vec<FilesystemEntity<String>>,
}

impl From<&ImportableCorpus> for ImportCorpus {
    fn from(importable_corpus: &ImportableCorpus) -> ImportCorpus {
        ImportCorpus {
            file_name: importable_corpus.file_name.clone(),
            format: importable_corpus.format,
            trace: importable_corpus.trace.clone(),
        }
    }
}

#[derive(Clone, Debug, Serialize)]
#[serde(tag = "type")]
#[serde(rename_all = "snake_case")]
#[serde(rename_all_fields = "camelCase")]
pub enum ImportCorpusResult {
    Imported { corpus: ImportedCorpus },
    Failed { message: String },
}
