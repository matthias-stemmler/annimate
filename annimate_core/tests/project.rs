use std::fs;
use std::path::Path;

use annimate_core::{
    AnnoKey, Project, ProjectContext, ProjectExportColumn, ProjectExportFormat, QueryLanguage,
};
use serde::Serialize;

const OUTPUT_DIR: &str = concat!(env!("CARGO_TARGET_TMPDIR"), "/tests/project/output");

macro_rules! project_test {
    ($(
        $name:ident: {
            corpus_set: $corpus_set:expr,
            corpus_names: $corpus_names:expr,
            aql_query: $aql_query:expr,
            query_language: $query_language:expr,
            export_columns: [$($export_column:expr),*$(,)?],
            export_format: $export_format:expr,
        }
    )*) => { $(
        #[test]
        fn $name() {
            let test_project = TestProject {
                corpus_set: $corpus_set,
                corpus_names: &$corpus_names,
                aql_query: $aql_query,
                query_language: {
                    use QueryLanguage::*;
                    $query_language
                },
                export_columns: {
                    #[allow(unused_imports)]
                    use TestProjectExportColumn::*;
                    #[allow(unused_imports)]
                    use TestProjectContext::*;
                    vec![$($export_column,)*]
                },
                export_format: {
                    use TestProjectExportFormat::*;
                    $export_format
                },
            };
            let project = test_project.clone().into();
            let project_debug = format!("{:?}", project);

            fs::create_dir_all(Path::new(OUTPUT_DIR)).unwrap();
            let project_file = Path::new(OUTPUT_DIR).join(concat!(stringify!($name), ".anmt"));
            let _ = fs::remove_file(&project_file);

            annimate_core::save_project(project, &project_file).unwrap();

            let output = fs::read_to_string(&project_file).unwrap();
            insta::with_settings!(
                {
                    info => &test_project,
                    omit_expression => true,
                },
                { insta::assert_snapshot!(output) }
            );

            let loaded_project = annimate_core::load_project(&project_file).unwrap();

            assert_eq!(project_debug, format!("{:?}", loaded_project));
        }
    )* };
}

project_test! {
    default: {
        corpus_set: None,
        corpus_names: [],
        aql_query: "",
        query_language: AQL,
        export_columns: [],
        export_format: Csv,
    }
    with_corpus_set: {
        corpus_set: Some("Test Corpus Set"),
        corpus_names: [],
        aql_query: "",
        query_language: AQL,
        export_columns: [],
        export_format: Csv,
    }
    with_one_corpus_name: {
        corpus_set: None,
        corpus_names: ["Test Corpus 1"],
        aql_query: "",
        query_language: AQL,
        export_columns: [],
        export_format: Csv,
    }
    with_two_corpus_names: {
        corpus_set: None,
        corpus_names: ["Test Corpus 1", "Test Corpus 2"],
        aql_query: "",
        query_language: AQL,
        export_columns: [],
        export_format: Csv,
    }
    with_aql_query: {
        corpus_set: None,
        corpus_names: [],
        aql_query: "Test AQL query",
        query_language: AQL,
        export_columns: [],
        export_format: Csv,
    }
    with_aql_query_complex: {
        corpus_set: None,
        corpus_names: [],
        aql_query: "a'''\"\nb",
        query_language: AQL,
        export_columns: [],
        export_format: Csv,
    }
    with_query_language_aql_quirks_v3: {
        corpus_set: None,
        corpus_names: [],
        aql_query: "",
        query_language: AQLQuirksV3,
        export_columns: [],
        export_format: Csv,
    }
    with_export_columns_default: {
        corpus_set: None,
        corpus_names: [],
        aql_query: "",
        query_language: AQL,
        export_columns: [
            Number,
            AnnoCorpus {
                anno_key: None,
            },
            AnnoDocument {
                anno_key: None,
            },
            AnnoMatch {
                anno_key: None,
                node_index: None,
            },
            MatchInContext {
                segmentation: None,
                context: Symmetric(20),
                primary_node_indices: &[],
            },
        ],
        export_format: Csv,
    }
    with_export_columns_with_annos: {
        corpus_set: None,
        corpus_names: [],
        aql_query: "",
        query_language: AQL,
        export_columns: [
            AnnoCorpus {
                anno_key: Some(("ns1", "anno1")),
            },
            AnnoDocument {
                anno_key: Some(("ns2", "anno2")),
            },
            AnnoMatch {
                anno_key: Some(("ns3", "anno3")),
                node_index: None,
            },
        ],
        export_format: Csv,
    }
    with_export_columns_with_node_indices: {
        corpus_set: None,
        corpus_names: [],
        aql_query: "",
        query_language: AQL,
        export_columns: [
            AnnoMatch {
                anno_key: None,
                node_index: Some(0),
            },
            MatchInContext {
                segmentation: None,
                context: Symmetric(20),
                primary_node_indices: &[1, 2, 3],
            },
        ],
        export_format: Csv,
    }
    with_export_columns_with_asymmetric_context: {
        corpus_set: None,
        corpus_names: [],
        aql_query: "",
        query_language: AQL,
        export_columns: [
            MatchInContext {
                segmentation: None,
                context: Asymmetric { left: 5, right: 10 },
                primary_node_indices: &[],
            },
        ],
        export_format: Csv,
    }
    with_export_columns_with_segmentation: {
        corpus_set: None,
        corpus_names: [],
        aql_query: "",
        query_language: AQL,
        export_columns: [
            MatchInContext {
                segmentation: Some("Test Segmentation"),
                context: Symmetric(20),
                primary_node_indices: &[],
            },
        ],
        export_format: Csv,
    }
    with_export_format_xlsx: {
        corpus_set: None,
        corpus_names: [],
        aql_query: "",
        query_language: AQL,
        export_columns: [],
        export_format: Xlsx,
    }
    full: {
        corpus_set: Some("Test Corpus Set"),
        corpus_names: ["Test Corpus 1", "Test Corpus 2"],
        aql_query: "Test AQL query",
        query_language: AQLQuirksV3,
        export_columns: [
            Number,
            AnnoCorpus {
                anno_key: Some(("ns1", "anno1")),
            },
            AnnoDocument {
                anno_key: Some(("ns2", "anno2")),
            },
            AnnoMatch {
                anno_key: Some(("ns3", "anno3")),
                node_index: Some(0),
            },
            MatchInContext {
                segmentation: Some("Test Segmentation"),
                context: Asymmetric { left: 5, right: 10 },
                primary_node_indices: &[1, 2, 3],
            },
        ],
        export_format: Xlsx,
    }
}

#[derive(Clone, Serialize)]
struct TestProject {
    corpus_set: Option<&'static str>,
    corpus_names: &'static [&'static str],
    aql_query: &'static str,
    query_language: QueryLanguage,
    export_columns: Vec<TestProjectExportColumn>,
    export_format: TestProjectExportFormat,
}

#[derive(Clone, Serialize)]
enum TestProjectExportColumn {
    Number,
    AnnoCorpus {
        anno_key: Option<(&'static str, &'static str)>,
    },
    AnnoDocument {
        anno_key: Option<(&'static str, &'static str)>,
    },
    AnnoMatch {
        anno_key: Option<(&'static str, &'static str)>,
        node_index: Option<u32>,
    },
    MatchInContext {
        segmentation: Option<&'static str>,
        context: TestProjectContext,
        primary_node_indices: &'static [u32],
    },
}

#[derive(Clone, Serialize)]
enum TestProjectContext {
    Symmetric(u32),
    Asymmetric { left: u32, right: u32 },
}

#[derive(Clone, Serialize)]
enum TestProjectExportFormat {
    Csv,
    Xlsx,
}

impl From<TestProject> for Project {
    fn from(test_project: TestProject) -> Self {
        Project {
            corpus_set: test_project.corpus_set.map(ToString::to_string),
            corpus_names: test_project
                .corpus_names
                .iter()
                .map(ToString::to_string)
                .collect(),
            aql_query: test_project.aql_query.to_string(),
            query_language: test_project.query_language,
            export_columns: test_project
                .export_columns
                .into_iter()
                .map(Into::into)
                .collect(),
            export_format: test_project.export_format.into(),
        }
    }
}

impl From<TestProjectExportColumn> for ProjectExportColumn {
    fn from(test_export_column: TestProjectExportColumn) -> Self {
        match test_export_column {
            TestProjectExportColumn::Number => ProjectExportColumn::Number,
            TestProjectExportColumn::AnnoCorpus { anno_key } => ProjectExportColumn::AnnoCorpus {
                anno_key: anno_key.map(|(ns, name)| AnnoKey {
                    ns: ns.into(),
                    name: name.into(),
                }),
            },
            TestProjectExportColumn::AnnoDocument { anno_key } => {
                ProjectExportColumn::AnnoDocument {
                    anno_key: anno_key.map(|(ns, name)| AnnoKey {
                        ns: ns.into(),
                        name: name.into(),
                    }),
                }
            }
            TestProjectExportColumn::AnnoMatch {
                anno_key,
                node_index,
            } => ProjectExportColumn::AnnoMatch {
                anno_key: anno_key.map(|(ns, name)| AnnoKey {
                    ns: ns.into(),
                    name: name.into(),
                }),
                node_index,
            },
            TestProjectExportColumn::MatchInContext {
                segmentation,
                context,
                primary_node_indices,
            } => ProjectExportColumn::MatchInContext {
                segmentation: segmentation.map(|s| s.to_string()),
                context: context.into(),
                primary_node_indices: primary_node_indices.into(),
            },
        }
    }
}

impl From<TestProjectContext> for ProjectContext {
    fn from(test_context: TestProjectContext) -> Self {
        match test_context {
            TestProjectContext::Symmetric(c) => ProjectContext::Symmetric(c),
            TestProjectContext::Asymmetric { left, right } => {
                ProjectContext::Asymmetric { left, right }
            }
        }
    }
}

impl From<TestProjectExportFormat> for ProjectExportFormat {
    fn from(test_export_format: TestProjectExportFormat) -> Self {
        match test_export_format {
            TestProjectExportFormat::Csv => ProjectExportFormat::Csv,
            TestProjectExportFormat::Xlsx => ProjectExportFormat::Xlsx,
        }
    }
}
