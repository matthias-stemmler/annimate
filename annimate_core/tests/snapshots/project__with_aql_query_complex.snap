---
source: annimate_core/tests/project.rs
info:
  corpus_set: ~
  corpus_names: []
  aql_query: "a'''\"\nb"
  query_language: AQL
  export_columns: []
  export_format: Csv
---
# Annimate project file
# https://github.com/matthias-stemmler/annimate

format-version = 1

[project]
query = """
a'''\"
b"""
query-language = "aql"
export-format = "csv"
